pkg/storage/zdb.go

```diff
@@ -36,116 +70,120 @@ func (s *storageModule) Allocate(diskType pkg.DeviceType, size uint64, mode pkg.
-		}
-
-		// Make sure adding this filesystem would not bring us over the disk limit
-		if reserved+size > usage.Size {
-			slog.Info().Msgf("Disk does not have enough space left to hold filesystem")
-			continue
-		}
-
 		volumes, err := pool.Volumes()
 		if err != nil {
-			return "", "", errors.Wrapf(err, "failed to list volume on pool %s", pool.Name())
+			return allocation, errors.Wrapf(err, "failed to list volume on pool %s", pool.Name())
 		}
 
 		for _, volume := range volumes {
-
 			// skip all non-zdb volume
-			if !strings.HasPrefix(volume.Name(), zdbPoolPrefix) {
+			if !isZdbVolume(volume) {
 				continue
 			}
 
-			usage, err := volume.Usage()
-			if err != nil {
-				return "", "", errors.Wrapf(err, "failed to read usage of volume %s", volume.Name())
+			zdb := zdbpool.New(volume.Path())
+
+			if !zdb.Exists(nsID) {
+				continue
 			}
 
-			// existing volume with enough storage, grow its limit
-			if err := volume.Limit(usage.Size + size); err != nil {
-				return "", "", errors.Wrapf(err, "failed to grow limit of volume %s", volume.Name())
+			// we found the namespace
+			allocation = pkg.Allocation{
+				VolumeID:   volume.Name(),
+				VolumePath: volume.Path(),
 			}
 
-			slog.Info().
-				Str("volume", volume.Name()).
-				Str("path", volume.Path()).
-				Msg("space allocated")
-			return volume.Name(), volume.Path(), nil
+			return allocation, nil
 		}
 	}
 
-	name, err := genZDBPoolName()
-	if err != nil {
-		return "", "", errors.Wrap(err, "failed to generate new sub-volume name")
+	type Candidate struct {
+		filesystem.Volume
+		Free uint64
 	}
 
-	volume, err := s.createSubvol(size, name, diskType)
-	if err != nil {
-		return "", "", errors.Wrap(err, "failed to create sub-volume")
-	}
-
-	slog.Info().
-		Str("volume", volume.Name()).
-		Str("path", volume.Path()).
-		Msg("space allocated")
-	return volume.Name(), volume.Path(), nil
-}
-
-// Claim let the system claim the allocated storage used by a 0-db namespace
-func (s *storageModule) Claim(name string, size uint64) error {
-	var (
-		v   filesystem.Volume
-		err error
-	)
-
+	var candidates []Candidate
+	// okay, so this is a new allocation
+	// we search all the pools for a zdb subvolume that can
+	// hold the new namespace
 	for _, pool := range s.volumes {
+		// skip pool with wrong disk type
+		if pool.Type() != diskType {
+			continue
+		}
+
+		usage, err := pool.Usage()
+		if err != nil {
+			return allocation, errors.Wrapf(err, "failed to read usage of pool %s", pool.Name())
+		}
 
 		volumes, err := pool.Volumes()
 		if err != nil {
-			return errors.Wrapf(err, "failed to list volume on pool %s", pool.Name())
+			return allocation, errors.Wrapf(err, "failed to list volume on pool %s", pool.Name())
 		}
 
 		for _, volume := range volumes {
-			if volume.Name() == name {
-				v = volume
-				break
+			// skip all non-zdb volume
+			if !isZdbVolume(volume) {
+				continue
+			}
+
+			zdb := zdbpool.New(volume.Path())
+
+			reserved, err := zdb.Reserved()
```

Comment:
Here again the way we compute the reserved storage of a disk is not correct. In this case we only check the amount of storage used by 0-db namespaces but we omit the storage reserved by container volumes.

Reply:
The 0-db subvolume is created with "unlimited" space (size == 0). Hence the only way to compute this is by using the zdbpool reserved.

Yesterday I was considering actually make the btrfs a bit smarter, and that they use the zdbpool to compute the "reserved" space of a subvolume if it is a zdb subvolume. In that case even subvolume creations will take namespaces reservations into account. What do u think? should we do that?

Reply:
yes we should